crates/client-api/src/routes/subscribe.rs

```diff
@@ -181,356 +194,631 @@ where
+                    | WsError::Utf8
+                    | WsError::AttackAttempt
+                    | WsError::Url(_)
+                    | WsError::Http(_)
+                    | WsError::HttpFormat(_)) => {
+                        log::warn!("Websocket receive error: {e}");
+                        break;
+                    }
+                },
             }
-            Item::HandleResult(res) => {
-                if let Err(e) = res {
-                    if let MessageHandleError::Execution(err) = e {
-                        log::error!("reducer execution error: {err:#}");
-                        // Serialize the message and keep a handle to the buffer.
-                        let (msg_alloc, msg_data) = serialize(msg_buffer, err, client.config);
-
-                        let send = async { ws.send(datamsg_to_wsmsg(msg_data)).await };
-                        let send = tokio::time::timeout(SEND_TIMEOUT, send);
-
-                        match send.await {
-                            Ok(Err(error)) => {
-                                log::warn!("Websocket send error: {error}")
-                            }
-                            Err(error) => {
-                                log::warn!("send timed out after: {error}");
-                                break;
-                            }
-                            _ => {}
-                        }
-
-                        // At this point,
-                        // the underlying allocation of `msg_data` should have a single referent
-                        // and this should be `msg_alloc`.
-                        // We can put this back into our pool.
-                        msg_buffer = msg_alloc
-                            .try_reclaim()
-                            .expect("should have a unique referent to `msg_alloc`");
+        }
+    }
+}
 
-                        continue;
+/// Stream that consumes [`ClientMessage`]s and yields [`DataMessage`]s for
+/// evaluation.
+///
+/// Calls `state.set_ponged()` if and when the input yields a pong message.
+/// Calls `state.close()` if and when the input yields a close frame,
+/// i.e. the client initiated a close handshake, which we track using the
+/// `client_closed_metric`.
+///
+/// Terminates if and when the input stream terminates.
+fn ws_client_message_handler(
+    state: Arc<ActorState>,
+    client_closed_metric: IntGauge,
+    mut messages: impl Stream<Item = ClientMessage> + Unpin,
+) -> impl Stream<Item = (DataMessage, Instant)> {
+    stream! {
+        while let Some(message) = messages.next().await {
+            match message {
+                ClientMessage::Message(message) => {
+                    log::trace!("Received client message");
+                    yield (message, Instant::now());
+                },
+                ClientMessage::Ping(_bytes) => {
+                    log::trace!("Received ping from client {}", state.client_id);
+                },
```

Comment:
Do we need or want to send a `Pong` response here? Does Tungstenite handle that automatically? If so, a comment to that effect would be appreciated.